src/librustc_mir/interpret/eval_context.rs

```diff
@@ -881,47 +878,77 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         self.raw_const_to_mplace(val)
     }
 
-    pub fn dump_place(&self, place: Place<M::PointerTag>) {
-        // Debug output
-        if !log_enabled!(::log::Level::Trace) {
-            return;
+    #[must_use]
+    pub fn dump_place(&'a self, place: Place<M::PointerTag>) -> PlacePrinter<'a, 'mir, 'tcx, M> {
```

Comment:
This is a neat pattern. Can we encapsulate this more by returning `impl Display` and not even exposing `PlacePrinter`, i.e., keeping that type inside the function?

Reply:
we likely can, but I basically gave up because the where bounds on the `impl Debug`'s lifetimes made me go crazy. I think the problem is that loads of our `impl` blocks don't have the `'mir: 'tcx` bound, so they can't call functions from impl blocks without that.